### ISSUE TITLE
Enabling efficient model-level redistribution between FSPD-TP

### ISSUE THREAD

**Issue by salmanmohammadi** [pytorch/pytorch#146601]
### 🚀 The feature, motivation and pitch

The age of RL-based LLM finetuning is upon is. For many RL training paradigms, there exists a step where a (large) model is used for inference (i.e. autoregressive sampling under `no_grad`), followed by a training step with the same model. After this training step, the updated model weights are used once more for inference, and so on. This generation step is often a large bottleneck in the overall training procedure.

For single-device RL finetuning, this is fairly straightforward to accomplish. However, in distributed training settings, we would like to optimally parallelize the model for inference using e.g. PP/TP, whilst for training we'd like to leverage FSDP to train large models. In the ideal setting, we would like zero-redundancy - we don't wan't to store two copies of the model at once, and we'd like this switch between inference-and-training modes to involve minimal overhead.

I've considered a few options:
- Gather the full state-dict on CPU, and re-initialize a new model instance with the desired parallel strategy. This seems prohibitively slow for large models, though I'm definitely open to suggestions as this may be the path of least resistance.
- Use `DTensor.redistribute` on the [weights of a model](https://github.com/SalmanMohammadi/torch-redistribute/blob/main/torch_redistribute/style.py#L40) sharded with `fully_shard` *and*:
  - Supporting some un-wrapping API for a model sharded with `fully_shard` - temporarily removing the pre/post-forward hooks, and then re-sharding the model weights to the original desired placements. This would allow us to use the same model instance. Alternatively,
  - Re-sharding the weights of a model sharded with `fully_shard` directly into another instance of the model initialized on meta device. I believe [veRL](https://github.com/volcengine/verl) may do something similar here (cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @ppwwyyxx @tianyu-l @XilunWu @SherlockNoMad @eric-haibin-lin), but this results in 2x parameters in memory unless parameters are offloaded for both resharding operations (FSDP->TP, and TP->FSDP).

Any advice on the best way to accomplish this would be appreciated. Thanks for reading : ) 

(cc @weifengpy @ebsmothers @joecummings @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o )

**Comment by wconstab** [comment]
to confirm I understand the goal,

you want to transform the model from (1) pure FSDP sharding, to pure TP sharding, (2) do inference, then (3) transform back to FSDP, and do so as efficiently as possible without adding extra memory pressure.  note: i'd assume we'd need at least a duplicate temporary buffer for one parameter at a time while doing these transforms?

RE efficiency, 
* obviously we'd want to minimize the number of communication steps needed (e.g. if we can do one alltoall that might be better than doing an 'fsdp unshard' followed by a 'dtensor redistribute').  
* it would be even better if we could hide the communication latency behind some compute.  Do you see such an opportunity, or did you assume the comm time would be fully exposed?

As far as implementation, it seems like it should be pretty easy to construct some API or context manager that accomplishes these transformation on the model but i'd want to understand the comms and the desired overlapping first.

**Comment by weifengpy** [comment]
I learned a lot from the proposal! thanks for doing this

> Gather the full state-dict on CPU, and re-initialize a new model instance with the desired parallel strategy

does this assume full state-dict fits on cpu memory? 

> Supporting some un-wrapping API for a model sharded with fully_shard - temporarily removing the pre/post-forward hooks, and then re-sharding the model weights to the original desired placements

is the "original deseried placements" referring to TP?  

> this results in 2x parameters in memory unless parameters are offloaded for both resharding operations (FSDP->TP, and TP->FSDP)

ideally the memory could be reduced from 2x parameters to 1x paramter + 1 buffer for redistribution?

I am not thinking of any compute-communication overlapping yet

**Comment by fegin** [comment]
@SalmanMohammadi 
> Supporting some un-wrapping API for a model sharded with fully_shard - temporarily removing the pre/post-forward hooks, and then re-sharding the model weights to the original desired placements. This would allow us to use the same model instance. Alternatively,

We probably don't need to modify `fully_shard`. We can just use `model.named_parameters()`, which will return a set of DTensors and use the local tensors of DTensors to reconstruct another model, using the `register_parameter`, similar to how FSDP1 to unregister/register parameters. 

@wconstab 
> obviously we'd want to minimize the number of communication steps needed (e.g. if we can do one alltoall that might be better than doing an 'fsdp unshard' followed by a 'dtensor redistribute').

hmm, why do we need alltoall? If it is FSDP -> TP, then one unshard following a local chunk will be sufficient.

**Comment by SalmanMohammadi** [comment]
Thank you for the responses @wconstab @weifengpy @fegin.

> does this assume full state-dict fits on cpu memory?

Yes, though I think this is a safe assumption given the compute capabilities most users with access to distributed setups have. Regardless, I think this operation would require a lot of communication and result in idle GPUs.

> is the "original deseried placements" referring to TP?

In this case the order of operations I had in mind was FSDP (shard on dim 0) -> TP (redistribute w/parallelize plan) -> FSDP (shard on dim 0, the original desired placements).

> ideally the memory could be reduced from 2x parameters to 1x paramter + 1 buffer for redistribution?

IIUC, do you mean 1x parameter + 1 buffer for each parameter, which is freed after that parameter has been redistributed?

> We probably don't need to modify fully_shard. We can just use model.named_parameters(), which will return a set of DTensors and use the local tensors of DTensors to reconstruct another model, using the register_parameter, similar to how FSDP1 to unregister/register parameters.

Thanks for pointing this API out - I wasn't aware of it. I'll try rework my hacky redistribution code to use this.

**Comment by janeyx99** [comment]
Totally late to the party, but if I'm understanding correctly that we're going FSDP training 1 -> TP inference 1 -> FSDP training 2 -> inference 2 -> training 3 ..., I'd expect that the weights from the inference portions won't change, and thus can be loaded asynchronously as the inference is happening for the next step. So in theory would it be correct to say that while inference i is happening, weights for training i + 1 can be loaded?

This will also only save memory if we can start dropping the first layers in TP as the computation progresses, which I am not confident enough about how TP works to say.

**Comment by SalmanMohammadi** [comment]
> Totally late to the party, but if I'm understanding correctly that we're going FSDP training 1 -> TP inference 1 -> FSDP training 2 -> inference 2 -> training 3 ..., I'd expect that the weights from the inference portions won't change, and thus can be loaded asynchronously as the inference is happening for the next step. So in theory would it be correct to say that while inference i is happening, weights for training i + 1 can be loaded?

Correct, the weights **at the end** of $FSDP_i$, $TP_i$, and **at the start** of $FSDP_{(i+1)}$ will all be identical. $FSDP_i$ and $FSDP_{(i+1)}$ will also be sharded identically. 

> So in theory would it be correct to say that while inference i is happening, weights for training i + 1 can be loaded?

I think so. The only complication here is that for the "inference" step, you're typically generating a completion given a prompt - one token (forward pass) at a time. I realize "inference" may have implied a single forward pass. In some cases (e.g. GRPO) you're generating multiple sequences - iteratively, not batched (i.e. `for i in n_rollouts: model.generate(prompt)...`. My point being is that it would be tricky to reliably know when the last step of inference is happening for you to overlap the computation of the final forward pass with re-sharding to FSDP.

> This will also only save memory if we can start dropping the first layers in TP as the computation progresses, which I am not confident enough about how TP works to say.

I roughly feel like if you followed something like [torchtitan's](https://github.com/pytorch/torchtitan/blob/49c6d6fc15ef644e5c3b1003ad4e0d9ea5fcb9a9/torchtitan/parallelisms/parallelize_llama.py#L113) TP parallelize plan this would be possible.

**Comment by fegin** [comment]
Another question, is the FSDP degree the same as TP degree? Mainly are you using the same number of GPUs for FSDP and for TP, no other parallelisms involve?

**Comment by SalmanMohammadi** [comment]
> are you using the same number of GPUs for FSDP and for TP

I would be surprised if there was a use-case to re-shard the same weights across *different* devices - I haven't seen one yet. I feel like one would be better off managing a different model instance and just syncing the weights in that case.

> no other parallelisms involve[d]?

I'd reccomend checking out [this paper](https://arxiv.org/pdf/2409.19256v2) (section 5.2.). A relevant quote:

> To eliminate redundant actor model copies, we advocate actor training and generation stages on the same of devices, 𝑁𝑎 GPUs allocated to the actor, and execute sequentially on the same copy of actor model weights. Nonetheless, actor training and generation may well adopt different 3D parallelism strategies, i.e., the generation stage typically requires smaller TP and PP sizes but a larger DP size, than the training stage.

They illustrate this in Figure 8 of the paper: 

<img width="619" alt="Image" src="https://github.com/user-attachments/assets/a23275f4-65d8-4bc5-a9e5-b7e7add150dc" />

Where the sets of "inference" (generation) parallelism and training parallelism strategies may be partially overlapping, and an optimal re-sharding solution could leverage this to avoid redundant communication.
